### PR TITLE
Support components multiple implementation for package name

### PR DESCRIPTION
**Pull request by GustavoCaso**
### What does this PR do?

Allow naming the component implementation package different base on the implementation name

### Motivation

Avoid package name duplication between component implementations

### Describe how to test/QA your changes

No code changes

### CODE REVIEW

**Review by davidor — APPROVED** [review]
:+1: for container-platform files

**Review by kayayarai — APPROVED** [review]
docs updates look good

**Review by louis-cqrl — APPROVED** [review]
LGTM

**Review by ichinaski — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]

**Review by vickenty — APPROVED** [review]

File: docs/public/components/creating-components.md [hunk 1]
```diff
@@ -114,11 +114,11 @@ internally by each component (more on this [here TODO]()). -->
 ### The impl folders
 
 The `impl` folder is where the component implementation is written. The details of component implementation are up to the developer.
-The only requirements are that the package name follows the pattern `<COMPONENT_NAME>impl` and that there is a public instantiation function called `NewComponent`.
+The only requirements are that the package name follows the pattern `<COMPONENT_NAME>impl` for the regular implementation or `<IMPL_NAME>impl` for the alternative implementation, and that there is a public instantiation function called `NewComponent`.
```
Comment (misteriaud) [inline docs/public/components/creating-components.md:117]: **💬 suggestion:** ‏Maybe we should write down that the directory name have to be format as follow `impl-<IMPL_NAME>` in accordance with the rule you add in the linter.

**Review by dustmop — APPROVED** [review]
To add some context here, the original decision was to require a component's alternative implementations to all use the same package name. The rationale was:

1) Only one implementation should be in use at a time, so there is no need to differentiate them.
2) From the perspective of the implementation, it is independent and unaware of the others, so it is *the* only implementation that exists.

This criteria was fine for the time, as it was focusing on alternatives that differ by their feature set. For example, a compressor using zlib vs a compressor using some other (or a trivial) compression method. But this PR shines light on a different use case: implementations that serve as being purpose-made for a particular build target. For this case, I agree that it makes sense for the implementation to talk about itself in relation to the binary that it expects to exist within, for example using `systemprobeimpl` for an implementation meant for the system-problem. This change also impacts implementations that differ by feature set, but in reevaluating the original rationale, I don't think it's worth the complexity of trying to maintain the old way. Said another way, I support relaxing the original restriction and going with this version instead.

**Review by jmw51798 — APPROVED** [review]

**Review by jackgopack4 — APPROVED** [review]

**Review by tobz — APPROVED** [review]

**Pull request merged**